src/common/databaseFilters.ts

```diff
@@ -3,57 +3,86 @@ import { displayPostcodeForHomelessClient } from "./format";
+const textFilterDelimiter = ",";
+
+export const dbFilterWithSubstringQueries = <DbData extends DbClientRow | DbParcelRow>(
+    substringToSubqueryMap: (value: string) => string
 ): ServerSideFilterMethod<DbData, string> => {
     return (query, state) => {
-        const cleanState = state.replace(/[^a-zA-Z0-9 ]/g, "");
-        if (cleanState === "") {
+        const substrings = state
+            .split(textFilterDelimiter)
+            .map((substring) => substring.trim().replace(/[^a-zA-Z0-9 \-+?]/g, ""))
+            .filter((substring) => substring.length > 0);
+
+        if (substrings.length === 0) {
             return query;
         }
-        if (parcelsPageDeletedClientDisplayName.toLowerCase().includes(cleanState.toLowerCase())) {
-            return query.or(
-                `${fullNameColumnLabel}.ilike.%${cleanState}%, ${clientIsActiveColumnLabel}.eq.false`
-            );
-        }
-        return query.ilike(fullNameColumnLabel, `%${cleanState}%`);
+
+        return query.or(substrings.map(substringToSubqueryMap).join(","));
     };
 };
 
+export const fullNameSearch = <DbData extends DbClientRow | DbParcelRow>(
+    fullNameColumnLabel: Extract<keyof DbData, "full_name" | "client_full_name">,
+    clientIsActiveColumnLabel: Extract<keyof DbData, "is_active" | "client_is_active">
+): ServerSideFilterMethod<DbData, string> => {
+    return dbFilterWithSubstringQueries((substring) => {
+        if (parcelsPageDeletedClientDisplayName.toLowerCase().includes(substring.toLowerCase())) {
+            return `or(${fullNameColumnLabel}.ilike.%${substring}%, ${clientIsActiveColumnLabel}.is.false)`;
+        }
+        return `and(${clientIsActiveColumnLabel}.is.true, ${fullNameColumnLabel}.ilike.%${substring}%)`;
+    });
+};
+
 export const postcodeSearch = <DbData extends DbClientRow | DbParcelRow>(
     postcodeColumnLabel: Extract<keyof DbData, "address_postcode" | "client_address_postcode">,
     clientIsActiveColumnLabel: Extract<keyof DbData, "is_active" | "client_is_active">
 ): ServerSideFilterMethod<DbData, string> => {
-    return (query, state) => {
-        if (state === "") {
-            return query;
-        }
-        if (state === "-") {
-            return query.is(clientIsActiveColumnLabel, false);
+    return dbFilterWithSubstringQueries((substring) => {
+        if (substring === "-") {
+            return `${clientIsActiveColumnLabel}.is.false`;
         }
-        if (displayPostcodeForHomelessClient.toLowerCase().includes(state.toLowerCase())) {
-            return query
-                .or(`${postcodeColumnLabel}.ilike.%${state}%, ${postcodeColumnLabel}.is.null`)
-                .neq(clientIsActiveColumnLabel, false);
+        if (displayPostcodeForHomelessClient.toLowerCase().includes(substring.toLowerCase())) {
+            return `and(${clientIsActiveColumnLabel}.is.true, or(${postcodeColumnLabel}.ilike.%${substring}%, ${postcodeColumnLabel}.is.null))`;
```

Comment:
I'm not thrilled that the subqueries are all written as text, but I spent a while trying to find a more elegant way without luck.

Reply:
I've seen this in the supabase docs somewhere, it is the "correct" way to do it. I don't love it either, but I'm pretty sure there isn't a better way